README.md
# Animated Progress Button
![](https://teamview-server.herokuapp.com/images/silas-transinghia.png)


Animated loading button to make you app button looks smoother.
Take inspritation from Pinterest https://www.pinterest.com/pin/671177150707141229/


![alt text](https://teamview-server.herokuapp.com/images/animated_progress_button/illustration.gif)

## Usages

Create `animatedButtonController` for controlling the button.

```sh
import 'package:animated_progress_button/animated_progress_button.dart';

final AnimatedButtonController animatedButtonController = AnimatedButtonController();
```

Then customize the button like your own way
*Noted that you must call `animatedButtonController.completed()` after getting response from server in order to complete the button animation*

 ```sh
AnimatedButton(
    controller: animatedButtonController,
    color: Colors.greenAccent,
    text: 'Log in',
    loadingText: 'Loading',
    loadedIcon: Icon(Icons.check, color: Colors.white),
    onPressed: () async {
        /// calling your API here and wait for the response.
        await Future.delayed(Duration(seconds: 5)); // simulated your API requesting time.
        animatedButtonController.completed(); // call when you get the response
        await Future.delayed(Duration(seconds: 2));
        animatedButtonController.reset(); // call to reset button animation
    },
),
```


CHANGELOG.md
## [1.0.0] - Release first version about animated button.

## [1.0.1] - Update package information.

## [1.0.2] - Add example

## [1.0.2+1] - Formatted according to dartfmt

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutteranimationplayground/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/animated_progress_button.dart
library animated_progress_button;

import 'package:flutter/material.dart';

class AnimatedButton extends StatefulWidget {
  const AnimatedButton(
      {Key key,
      @required this.color,
      @required this.text,
      @required this.controller,
      this.loadedIcon,
      this.height,
      this.loadingText,
      this.onPressed})
      : super(key: key);

  /// Color of the button
  final Color color;

  /// Initial text display
  final String text;

  /// Loading text
  final String loadingText;

  /// You can also customized the last icon appear in the animation progress.
  final Widget loadedIcon;

  /// Default 50.
  final double height;

  /// AnimatedButtonController to help control the animation progress.
  /// Call [controller.completed()] when getting response from your request
  /// Call [controller.reset()] to restart button animation
  final AnimatedButtonController controller;

  /// Customize on pressing action
  final VoidCallback onPressed;

  @override
  _AnimatedButtonState createState() => _AnimatedButtonState();
}

class _AnimatedButtonState extends State<AnimatedButton>
    with TickerProviderStateMixin {
  static int simulatedLoadingTime = 3;
  bool _isLoading = false;
  AnimationController _animationController;
  Animation<double> _progressAnimation;
  Animation<double> _buttonWidthAnimation;
  Animation<Offset> _checkIconOffsetAnimation;

  @override
  void initState() {
    super.initState();
    widget.controller?.addLoadedListener((bool shouldReset) {
      if (shouldReset) {
        _animationController.reset();
        _isLoading = false;
        return;
      }
      _animationController.forward(from: 0.6);
    });
    _animationController = AnimationController(
        duration: Duration(seconds: simulatedLoadingTime), vsync: this)
      ..addListener(() {
        if (_animationController.value >= 0.6 &&
            !(widget.controller?.value ?? true)) {
          _animationController.stop(canceled: false);
        }
        setState(() {});
      });
    _progressAnimation = Tween<double>(begin: 0.0, end: 1.0).animate(
        CurvedAnimation(
            parent: _animationController,
            curve: Interval(0.0, 0.7, curve: Curves.ease)));
    _buttonWidthAnimation = Tween<double>(begin: 400, end: 50).animate(
        CurvedAnimation(
            parent: _animationController,
            curve: Interval(0.7, 0.9, curve: Curves.decelerate)));
    _checkIconOffsetAnimation =
        Tween<Offset>(begin: Offset(0, -2), end: Offset.zero).animate(
            CurvedAnimation(
                parent: _animationController,
                curve: Interval(0.9, 1.0, curve: Curves.easeOutBack)));
  }

  bool get _progressLoadingCompleted => _progressAnimation.value == 1.0;
  bool get _visibleIcon => _checkIconOffsetAnimation.value.dy > -2;

  @override
  Widget build(BuildContext context) {
    return Container(
      height: widget.height ?? 50,
      width: _buttonWidthAnimation.value,
      margin: EdgeInsets.only(top: 32.0, left: 28.0, right: 28.0),
      child: FlatButton(
        padding: EdgeInsets.zero,
        onPressed: () {
          if (_isLoading) return;
          _isLoading = true;
          _animationController.forward();
          if (widget.onPressed != null) {
            widget.onPressed();
          }
        },
        child: Container(
            decoration: BoxDecoration(
                borderRadius: BorderRadius.all(Radius.circular(32.0)),
                color: widget.color,
                boxShadow: [
                  BoxShadow(
                      offset: Offset(0, 3),
                      color: Colors.black12,
                      blurRadius: 10.0,
                      spreadRadius: 1.0)
                ]),
            child: Stack(
              children: <Widget>[
                Row(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: <Widget>[
                    Padding(
                      padding: const EdgeInsets.all(16.0),
                      child: Text(
                          _progressLoadingCompleted
                              ? ""
                              : _isLoading
                                  ? (widget.loadingText ?? 'In progress ...')
                                  : (widget.text ?? ''),
                          style: TextStyle(color: Colors.white, fontSize: 15)),
                    ),
                  ],
                ),
                _progressLoadingCompleted
                    ? Container()
                    : Container(
                        height: 50,
                        child: ClipRRect(
                          borderRadius: BorderRadius.all(Radius.circular(32.0)),
                          child: LinearProgressIndicator(
                            value: _progressAnimation.value,
                            valueColor: AlwaysStoppedAnimation<Color>(
                                Colors.white.withOpacity(.5)),
                            backgroundColor: Colors.transparent,
                          ),
                        ),
                      ),
                Visibility(
                  visible: _visibleIcon,
                  child: SlideTransition(
                      position: _checkIconOffsetAnimation,
                      child: Center(
                          child: widget.loadedIcon ??
                              Icon(Icons.check, color: Colors.white))),
                )
              ],
            )),
      ),
    );
  }
}

typedef void OnLoaded(bool shouldReset);

class AnimatedButtonController extends ValueNotifier<bool> {
  OnLoaded _callback;

  AnimatedButtonController({bool value = false}) : super(value);

  void completed() {
    value = true;
    if (_callback != null) {
      _callback(false);
    }
  }

  void addLoadedListener(OnLoaded callback) {
    this._callback = callback;
  }

  void reset() {
    value = false;
    if (_callback != null) {
      _callback(true);
    }
  }
}


example/lib/main.dart
import 'package:animated_progress_button/animated_progress_button.dart';
import 'package:flutter/material.dart';

Color mainColor = Color(0xFF41CEBB);

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      debugShowCheckedModeBanner: false,
      home: LoginPage(),
    );
  }
}

class LoginPage extends StatefulWidget {
  @override
  _LoginPageState createState() => _LoginPageState();
}

class _LoginPageState extends State<LoginPage> {
  final animatedButtonController = AnimatedButtonController();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Color(0xFFF4F6F4),
      appBar: AppBar(
        leading: BackButton(color: mainColor),
        backgroundColor: Colors.transparent,
        elevation: 0.0,
        brightness: Brightness.light,
        actions: <Widget>[FlatButton(onPressed: () {}, child: Text("Sign up"))],
      ),
      body: Column(
        children: <Widget>[
          Center(
              child: Padding(
            padding: const EdgeInsets.all(28.0),
            child: Text("Log in",
                style: TextStyle(fontWeight: FontWeight.bold, fontSize: 24)),
          )),
          Container(
            margin: EdgeInsets.only(top: 16.0),
            decoration: BoxDecoration(
                shape: BoxShape.circle,
                border: Border.all(color: Colors.white, width: 8.0),
                boxShadow: [
                  BoxShadow(
                      color: Colors.black12, blurRadius: 6.0, spreadRadius: 2.0)
                ]),
            child: ClipOval(
              child: Image.network(
                "https://iv1.lisimg.com/image/16128926/480full-isa-longwell.jpg",
                height: 100,
                width: 100,
                fit: BoxFit.cover,
              ),
            ),
          ),
          Container(
            margin: EdgeInsets.only(top: 56.0, left: 28.0, right: 28.0),
            decoration: BoxDecoration(
                color: Colors.white,
                borderRadius: BorderRadius.all(Radius.circular(32.0)),
                boxShadow: [
                  BoxShadow(
                      offset: Offset(0, 3),
                      color: Colors.black12,
                      blurRadius: 10.0,
                      spreadRadius: 1.0)
                ]),
            child: Center(
              child: Padding(
                padding: const EdgeInsets.only(left: 24.0, right: 24.0),
                child: TextField(
                  decoration: InputDecoration.collapsed(hintText: "Password"),
                  cursorColor: mainColor,
                  obscureText: true,
                ),
              ),
            ),
          ),
          AnimatedButton(
            color: mainColor,
            text: 'Log in',
            controller: animatedButtonController,
            onPressed: () async {
              /// calling your API here and wait for the response.
              await Future.delayed(
                  Duration(seconds: 5)); // simulated your API request.
              animatedButtonController.completed();
              await Future.delayed(Duration(seconds: 2));
              animatedButtonController.reset(); // call to reset button
            },
          ),
        ],
      ),
    );
  }
}
